### PR TITLE
Move [Span|TagMap]InScope and share between api/sdk

### LINKED ISSUE #333: Avoid duplicate code for Scope implementation
Currently in Default[Tracer|Tagger] we have an implementation of the Scope as well as in the SDK. I think this is not ideal and I would suggest that we remove the duplicated code.

I will suggest that we remove this duplicated code, the options that I see are:
1. We change Default[Tracer|Tagger] to not be final and [Tracer|Tagger]Sdk inherit from the Default.
1. We move the Scope implementation in the usafe.ContextUtil for every package so that both the Default and the Sdk implementation can share this.

Any thoughts?

### PR DESCRIPTION
**Pull request by tylerbenson**
Removes duplication between the two projects

Updates #333

### CODE REVIEW

File: api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * limitations under the License.
+ */
+
+package io.opentelemetry.tags.unsafe;
+
+import io.grpc.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.tags.TagMap;
+
+public final class TagMapInScope implements Scope {
```
Comment (songy23) [inline api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: Please add Javadoc.

File: api/src/main/java/io/opentelemetry/trace/unsafe/SpanInScope.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * limitations under the License.
+ */
+
+package io.opentelemetry.trace.unsafe;
+
+import io.grpc.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.Span;
+
+public final class SpanInScope implements Scope {
```
Comment (songy23) [inline api/src/main/java/io/opentelemetry/trace/unsafe/SpanInScope.java]: Same here.

File: api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * limitations under the License.
+ */
+
+package io.opentelemetry.tags.unsafe;
+
+import io.grpc.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.tags.TagMap;
+
+/** A scope that manages the {@link Context} for a {@link TagMap}. */
```
Comment (songy23) [inline api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: Nit: `@since 0.1.0`

File: api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+  private TagMapInScope(TagMap tags) {
+    orig = ContextUtils.withValue(tags).attach();
+  }
+
+  /**
+   * Constructs a new {@link TagMapInScope}.
+   *
+   * @param tags the {@code TagMap} to be added to the current {@code Context}.
+   */
```
Comment (songy23) [inline api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: ditto

File: api/src/main/java/io/opentelemetry/trace/unsafe/SpanInScope.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * limitations under the License.
+ */
+
+package io.opentelemetry.trace.unsafe;
+
+import io.grpc.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.Span;
+
+/** A scope that manages the {@link Context} for a {@link Span}. */
```
Comment (songy23) [inline api/src/main/java/io/opentelemetry/trace/unsafe/SpanInScope.java]: ditto

File: api/src/main/java/io/opentelemetry/trace/unsafe/SpanInScope.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  private SpanInScope(Span span) {
+    current = ContextUtils.withValue(span);
+    previous = current.attach();
+  }
+
+  /**
+   * Constructs a new {@link SpanInScope}.
+   *
+   * @param span the {@code Span} to be added to the current {@code Context}.
+   */
```
Comment (songy23) [inline api/src/main/java/io/opentelemetry/trace/unsafe/SpanInScope.java]: ditto

**Comment by tylerbenson** [comment]
@songy23 thanks for all the nits.  I'll remember to do them on my own eventually.

File: api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+   * @param tags the {@code TagMap} to be added to the current {@code Context}.
+   * @since 0.1.0
+   */
+  public static TagMapInScope create(TagMap tags) {
+    return new TagMapInScope(tags);
+  }
+
+  @Override
+  public void close() {
+    Context.current().detach(orig);
```
Comment (carlosalberto) [inline api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: Based on the `Context` documentation, we shouldn't do this, but save the context variable:

> If this method finds that this context is not current, either you or some code in-between are not detaching correctly, and a SEVERE message will be logged but the context to attach will still be bound. Never use Context.current().detach(), as this will compromise this error-detecting mechanism.
```

But maybe we should do that in another PR...

Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: Not sure I understand the comment. Can someone clarify?

File: api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import io.grpc.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.tags.TagMap;
+
+/**
+ * A scope that manages the {@link Context} for a {@link TagMap}.
+ *
+ * @since 0.1.0
+ */
+public final class TagMapInScope implements Scope {
```
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: Instead of exposing this class can we expose a method on ContextUtils:

```java
public Scope withTagMap(TagMap tagMap) {
...
}
```

Any thoughts on this approach?

File: api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java (hunk 7 above)
Comment (pavolloffay) [reply api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: The comment says we should not be using `Context.curret()` but rather store the current into the private field in `Scope`impl when the scope is created and at the close time call `current.detach(orig)`

Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: Let's create an issue and discuss about this, I will add the gRPC devs there as well.

File: api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java (hunk 8 above)
Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: That sounds good - that being said, I wished we had something more general-purpose, like `<T> Scope withValue(T value);`

File: api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java (hunk 7 above)
Comment (carlosalberto) [reply api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: Let's just open a (minor) Issue for this, so we don't get blocked. We can later update the code.

**Comment by tedsuo** [comment]
@tylerbenson just following up. Do @bogdandrutu and @carlosalberto comments make sense?

File: api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java (hunk 8 above)
Comment (bogdandrutu) [reply api/src/main/java/io/opentelemetry/tags/unsafe/TagMapInScope.java]: It is not possible to have this, because we need access to the Key in this module as well.
